Keep the thread pool small on startup for better large config startup performance

gjdeval opened the pull request:
Starting a large Liberty configuration (many large complex applications) can lead to contention for resources shared among the applications. Keeping the thread pool small during server/application startup can significantly reduce both startup duration and cpu used during startup. 

This PR sets a small startup pool size, which is fixed until server/application startup completes (or the timeout is reached). After startup is complete, the pool size is set to coreThreads and allowed to vary based on coreThreads/maxThreads tuning (if any).

gjdeval committed:
small thread pool on startup for better large config start performance

vijaysun-omr commented on dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java:
```diff
@@ -673,7 +682,7 @@ enum LastAction {
         this.maxThreads = pool.getMaximumPoolSize();
         this.threadRange = this.maxThreads - this.coreThreads;
-        setPoolSize(coreThreads);
+        setPoolSize(startupPoolSize);
```
Should there be any attempt to respect `maxThreads` during startup as well ? e.g. should `startupPoolSize` be allowed to override that potentially during startup (if it was a larger value than `maxThreads` that someone had for whatever reason) ?

gjdeval replied:
That's a good catch, thanks. I'll change that to 
`        setPoolSize(Math.min(startupPoolSize, this.maxThreads));`

Looking at that made me realize another issue ... 

Currently the thread pool controller does not run after the initial cycle if  `coreThreads == maxThreads` because there is nothing for it to do. But the addition of `startupPoolSize` means that the controller needs to run on its usual cycle until "ready to run", or it will never switch from `startupPoolSize` to `coreThreads`, so I need to change that also.

gjdeval replied:
OK, I've pushed an update to the PR that should fix both of the items mentioned above.

vijaysun-omr replied:
Thanks Gary, the update looks fine to me.

vijaysun-omr commented on dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java:
```diff
@@ -673,14 +682,11 @@ enum LastAction {
-            activeTask = new IntervalTask(this);
-            timer.schedule(activeTask, interval, interval);
-        }
+        activeTask = new IntervalTask(this);
```
I guess you could have done a test here for `coreThreads < maxThreads || startupPoolSize < maxThreads` but given that the default is different for `startupPoolSize` it is unlikely that you would end up not needing to run any controller logic. I assume the overhead of enabling this code is not expected to be significant even in the lightest (i.e. smallest task) known scenarios to us.

gjdeval replied:
I'm not sure I understand the question. 

When we only had `coreThreads` and `maxThreads`, if they were equal then the controller had nothing to do, so it did not need to wake up every 1500 ms to check the pool stats, so we did not set the timer.

Now that we are adding `startupPoolSize` to the mix, we need the controller to do the periodic checks until "ready to run" so that it can see the notification that startup is complete and switch from `startupPoolSize` to `coreThreads`. 

Those checks are extremely lightweight, just a couple of if statements in `evaluateInterval` (shown later in the changeset, starting on line 1344). But we have to do this, or the pool will stay in `startupPoolSize` forever ...

vijaysun-omr replied:
My question was for the case when `startupPoolSize == coreThreads == maxThreads`. The logic for not running the controller in that case is similar to why we would not have run it originally if `coreThreads == maxThreads`

vijaysun-omr replied:
The "overhead" I was asking about was for the controller doing its logic every 1.5 seconds on an ongoing basis even when there is nothing to do. If that overhead is negligible, then I agree there is  not (and never was) any point in avoiding running that logic (even originally when `coreThreads` == `maxThreads`)

gjdeval replied:
Ah-ha, now I understand the question, thanks. 

And yes, we could get a small optimization from not activating the timer in this case:
`startupPoolSize == coreThreads == maxThreads`

I can add that in there, because in that case there is _still_ nothing for the controller to do, so no point in it waking up every 1500 ms. The overhead is very small, but it will still add a bit to idle cpu time.

gjdeval replied:
Pushed a commit with the change to avoid running the controller cycle if `startupPoolSize == coreThreads == maxThreads`

vijaysun-omr replied:
Thanks, yes this is what I had in mind with my question.

gjdeval commented:
Functional correctness was tested by starting a large Liberty configuration with the various permutations of `startupPoolSize` and `coreThreads` and `maxThreads` with `ThreadPoolController` trace enabled and taking javacores during startup. With the trace.log and javacores, I observed that the controller cycle ran when it should have, and did not run when it should not have, and that the number of threads in the Liberty default pool was correct for the particular combination of settings used for that startup.

Performance benefit was evaluated by measuring startup time and cpu-secs used with a large Liberty configuration on two different lab systems with 42 and 48 cpus respectively available to the JVM, without and with a patch containing the changes in this PR. The patch reduced startup time by 10-20% and reduced cpu-secs used during startup by 40-45%. 

The regression patrol team will run the full suite of SUFP configurations with these changes and we will update the PR when those results are available.

gjdeval commented:
#build

jhanders34 committed:
Restructure com.ibm.ws.kernel.feature.core

- Pull out interfaces from com.ibm.ws.kernel.feature.core into
com.ibm.ws.kernel.feature.common to avoid circular dependencies with
runtime.update so that others can rely on the interfaces.

gjdeval commented:
Reworked the implementation based on discussion w/Jared, also added check for default startupPoolSize to unit test.

gjdeval commented:
#build

gjdeval commented:
#build

gjdeval commented:
Based on review comments, I have added a couple of refinements to the implementation:

- If the thread pool hangs during startup, the controller will switch out of startup mode to regular coreThreads/maxThreads operation, after which hang resolution will be invoked
- If the system property for startupPoolSize is set to -1 the controller will not use startupPoolSize, instead starting with the usual  coreThreads/maxThreads operation - this is a triage mechanism to work around unexpected problems

gjdeval commented:
#build

jhanders34 commented on dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java:
```diff
@@ -673,14 +701,34 @@ enum LastAction {
+                    activeTask = new IntervalTask(this);
+                    timer.schedule(activeTask, interval, interval);
+                }
+            } else {
+                // core == max so stop controller cycle if it is running
+                if (activeTask != null) {
+                    activeTask.cancel();
+                    activeTask = null;
+                }
```
```suggestion
```

Always is null since it isn't set in the constructor earlier.  No need for the logic.

gjdeval replied:
Good catch, thanks - made the change.

jhanders34 commented on dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java:
```diff
@@ -716,6 +764,19 @@ enum LastAction {
+     * Switch to regular pool sizing after startup completes
+     */
+    void startupCompleted() {
+        initialStartupCompleted = true;
```
Do we need to check to see if initialStartupComponent is false before doing this?  If we reset things in the hang scenario, we do not want to go back down to core thread size if we already expanded right?

gjdeval replied:
Good catch - made the change.

jhanders34 commented on dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java:
```diff
@@ -716,6 +764,19 @@ enum LastAction {
+        initialStartupCompleted = true;
+        setPoolSize(coreThreads);
+        // no need to run the controller cycle if there will be no pool size changes
+        if (coreThreads < maxThreads) {
+            activeTask = new IntervalTask(this);
+            timer.schedule(activeTask, interval, interval);
+        }
```
The task may already be running from the constructor.  Why starting it a second time?

gjdeval replied:
Good catch. I've added a check for (activeTask == null) before starting it up. 
Also added an else clause to stop the controller cycle if it was running during startup (because startupPoolSize < maxThreads) but should not run after startup if (coreThreads == maxThreads).

gjdeval commented:
#build

tjwatson commented on dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ExecutorServiceImpl.java:
```diff
@@ -75,6 +76,23 @@ public final class ExecutorServiceImpl implements WSExecutorService, ThreadQuies
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+            Tr.event(tc, ": server start complete.");
+        }
+        threadPoolController.startupCompleted();
```
What happens when a large number of applications are configured and take longer than the 30 seconds to start?  That defaults to registering the `ServerStarted` service before all the applications have really started.  If that has a very large set of applications to start will it then grow the pool to be "too big" and result in large contention again?

gjdeval replied:
The transition from "startup thread pool size" to "regular thread pool size" happens on `startupCompleted` so if `applicationManager startTimeout` fires before application start is complete, the remainder of application startup will use the regular thread pool sizing. The potential for `startTimeout` default of 30 seconds to be insufficient is preexisting and independent of these changes. There are already good reasons for users to configure `startTimeout` to allow enough time for all their applications to start, so that for example ports are not opened before the applications are ready to handle requests.

tjwatson replied:
Understood
